### PR TITLE
Plugin tmpl

### PR DESCRIPTION
**Pull request by yetist**
Update the plugin templates to generate new pluma plugin.

Usage:
```
$ ./tools/generate-plugin.py --help
Usage:
   generate-plugin.py [OPTIONS...] pluginname

generate skeleton source tree for a new pluma plugin.

Options:
  --author              Set the author name
  --email               Set the author email
  --description         Set the description you want for your new plugin
  --standalone          Is this plugin intended to be distributed as a
                        standalone package ? (N/A)
  --language / -l       Set the language (C) [default: c]
  --with-$feature       Enable $feature
  --without-$feature    Disable $feature
  --help / -h           Show this message and exits

Features:
  config-dlg            Plugin configuration dialog
  menu                  Plugin menu entries
  side-pane             Side pane item (N/A)
  bottom-pane           Bottom pane item (N/A)

```

### CODE REVIEW

File: tools/plugin_template/pluma-plugin.c [hunk 1]
```diff
@@ -25,160 +25,283 @@
...
 	manager = pluma_window_get_ui_manager (window);
 
 	data->action_group = gtk_action_group_new ("##(PLUGIN_ID.camel)PluginActions");
 	gtk_action_group_set_translation_domain (data->action_group,
 						 GETTEXT_PACKAGE);
 	gtk_action_group_add_actions (data->action_group,
-				      action_entries,
-				      G_N_ELEMENTS (action_entries),
-				      window);
+					   action_entries,
```
Comment (raveit65) [inline tools/plugin_template/pluma-plugin.c:161]: Indents.

File: tools/plugin_template/pluma-plugin.c [hunk 2]
```diff
@@ -25,160 +25,283 @@
...
-impl_create_configure_dialog (PlumaPlugin *plugin)
+##(PLUGIN_ID.lower)_plugin_create_configure_widget (PeasGtkConfigurable *configurable)
 {
 	pluma_debug (DEBUG_PLUGINS);
 }
 ##endif
 
+static void
+##(PLUGIN_ID.lower)_plugin_set_property (GObject      *object,
+                                guint         prop_id,
```
Comment (raveit65) [inline tools/plugin_template/pluma-plugin.c:220]: Indents

File: tools/plugin_template/pluma-plugin.c [hunk 3]
```diff
@@ -25,160 +25,283 @@
...
+			G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+			break;
+	}
+}
+
+static void
+##(PLUGIN_ID.lower)_plugin_get_property (GObject    *object,
+                                guint       prop_id,
+                                GValue     *value,
+                                GParamSpec *pspec)
```
Comment (raveit65) [inline tools/plugin_template/pluma-plugin.c:242]: Indents

**Comment by raveit65** [comment]
@mate-desktop/core-team 
Anyone else like to review it?

**Comment by mbkma** [comment]
if #615 gets merged, this will be out of date, because new plugins should use the new interfaces introduced in #615 .
So I would recommend to first review #615 and if it gets merged update this PR.

**Comment by raveit65** [comment]
ok, @yetist  ^^^^^

**Comment by raveit65** [comment]
@yetist 
Are you OK?

**Comment by lukefromdc** [comment]
Any progress on this?

**Comment by yetist** [comment]
Please review and test, for example:
```
cd plugins
# run generateplease-plugin.py to create newplugin, please run with -h to see more usage.
../tools/generate-plugin.py newplugin
# add newplugin into SUBDIRS and DIST_SUBDIRS in Makefile.am
vim Makefile.am
# add "plugins/newplugin/Makefile", "plugins/newplugin/newplugin.plugin.desktop.in" into  AC_CONFIG_FILES like others in configure.ac
vim ../configure.ac
cd ..
./autogen
make
```

**Comment by lukefromdc** [comment]
Someone who knows about the Pluma plugin system (I do not) needs to review this

File: tools/plugin_template/pluma-plugin.c [hunk 4]
```diff
@@ -25,160 +25,286 @@
...
+##ifdef WITH_CONFIGURE_DIALOG
+#include <libpeas-gtk/peas-gtk-configurable.h>
+##endif
 
-#define WINDOW_DATA_KEY	"##(PLUGIN_ID.camel)PluginWindowData"
+#include <pluma/pluma-window.h>
+#include <pluma/pluma-debug.h>
 
-#define ##(PLUGIN_ID.upper)_PLUGIN_GET_PRIVATE(object)	(G_TYPE_INSTANCE_GET_PRIVATE ((object), TYPE_##(PLUGIN_ID.upper)_PLUGIN, ##(PLUGIN_ID.camel)PluginPrivate))
+static void peas_activatable_iface_init (PeasActivatableInterface *iface);
```
Comment (mbkma) [inline tools/plugin_template/pluma-plugin.c:37]: We do not use `PeasActivatable`  anymore. Please see https://github.com/mate-desktop/pluma/blob/f628a6ef38b7c42febf8fbc450d46ab1f5461f6f/plugins/trailsave/pluma-trail-save-plugin.c#L31 for example for the current way to write a plugin.

File: tools/plugin_template/pluma-plugin.c [hunk 5]
```diff
@@ -25,160 +25,286 @@
...
+                                  (view != NULL) &&
+                                  gtk_text_view_get_editable (GTK_TEXT_VIEW (view)));
 ##endif
+}
+
 
 static void
-impl_activate (PlumaPlugin *plugin,
-	       PlumaWindow *window)
+##(PLUGIN_ID.lower)_plugin_activate (PeasActivatable *activatable)
```
Comment (mbkma) [inline tools/plugin_template/pluma-plugin.c:143]: Also this function needs to be updated, see https://github.com/mate-desktop/pluma/blob/f628a6ef38b7c42febf8fbc450d46ab1f5461f6f/plugins/trailsave/pluma-trail-save-plugin.c#L162 for example.

File: tools/plugin_template/pluma-plugin.c [hunk 6]
```diff
@@ -25,160 +25,286 @@
...
+		return;
+	}
+	update_ui (data);
 ##endif
 }
 
 static void
-impl_deactivate (PlumaPlugin *plugin,
-		 PlumaWindow *window)
+##(PLUGIN_ID.lower)_plugin_deactivate (PeasActivatable *activatable)
```
Comment (mbkma) [inline tools/plugin_template/pluma-plugin.c:184]: Also this function needs to be updated...

File: tools/plugin_template/pluma-plugin.c [hunk 7]
```diff
@@ -25,160 +25,286 @@
...
+                                         guint         prop_id,
+                                         const GValue *value,
+                                         GParamSpec   *pspec)
+{
+	##(PLUGIN_ID.camel)Plugin *plugin = ##(PLUGIN_ID.upper)_PLUGIN (object);
+
+	switch (prop_id)
+	{
+		case PROP_OBJECT:
+			plugin->priv->window = GTK_WIDGET (g_value_dup_object (value));
```
Comment (mbkma) [inline tools/plugin_template/pluma-plugin.c:232]: Also `plugin->priv->window` is now of type `PlumaWindow`.

**Review by mbkma — COMMENTED** [review]
In general you can see the plugins in the pluma repo for reference or the plugins from the repo https://github.com/mate-desktop/pluma-plugins

**Review by mbkma — APPROVED** [review]
Thanks, looks good now.

**Comment by lukefromdc** [comment]
We have one current review, this has been updated after cooking for two years. Is this ready to go? I don't know enough about the plugin system to evaluate it

**Comment by mbkma** [comment]
Build and runs successfully, thanks :)

**Pull request merged**